### PR TITLE
BTM-321: helpers for mocked invoices

### PR DESCRIPTION
**Pull request by gdsgj**
This adds code for integration tests that can help to create matching invoices and transactions.

Specifically, this adds one function and updates two others:
- `getVendorServiceConfigRow`: added to fetch vendor-service config from the private repo
- `randomLineItem` and `randomInvoice`: updated to allow some non-random fields

Here is an example of how to create an invoice and transactions with matching vendor, transaction quantity, and event names:
```ts
import {
  makeMockInvoicePDF,
  randomInvoice,
  randomLineItem,
  writeInvoiceToS3,
} from "../helpers/mock-data/invoice";
import {
  generatePublishAndValidateEvents,
  TimeStamps,
} from "../helpers/commonHelpers";
import { getVendorServiceConfigRow } from "../helpers/configHelpers";
import { ClientId, EventName } from "../payloads/snsEventPayload";

describe("testing", () => {
  test("testing", async () => {
    const givenClientId = ClientId.client1;
    const givenEventName = EventName.IPV_PASSPORT_CRI_REQUEST_SENT;
    const givenQuantity = 10;

    const { service_name: givenServiceName, vendor_name: givenVendorName } =
      await getVendorServiceConfigRow({
        client_id: givenClientId,
        event_name: givenEventName,
      });

    const givenLineItem = randomLineItem({
      description: givenServiceName,
      quantity: givenQuantity,
    });

    const givenInvoice = randomInvoice({
      vendor: {
        name: givenVendorName,
      },
      lineItems: [givenLineItem],
    });

    await Promise.all([
      generatePublishAndValidateEvents({
        numberOfTestEvents: givenQuantity,
        eventName: givenEventName,
        clientId: givenClientId,
        eventTime: TimeStamps.CURRENT_TIME,
      }),
      makeMockInvoicePDF(writeInvoiceToS3)(givenInvoice),
    ]);

    // Check the results
  });
});
```

### CODE REVIEW

File: integration_tests/helpers/configHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        fieldValue === undefined ||
+        row[fieldName as keyof VendorServiceConfigRow] === fieldValue
+    )
+  );
+
+  if (vendorServiceConfigRow === undefined)
+    throw Error("Vendor service config row not found");
+
+  return vendorServiceConfigRow;
+};
```
Comment (Si3rr4wow) [inline integration_tests/helpers/configHelpers.ts:52]: Just a suggestion, not necessary, but you could split this out to reduce nesting and avoid reassigning `cachedVendorServiceConfig` like this 
```
const cachedVendorServiceConfig = (async (): Promise<
  VendorServiceConfigRow[]
> => {
  const vendorServiceConfigText = await getS3Object({
    bucket: configStackName(),
    key: "vendor_services/vendor-services.csv",
  });

  if (vendorServiceConfigText === undefined)
    throw Error("No vendor service config found");

  const csvConverter = getCsvConverter();

  const vendorServiceConfig = await csvConverter.fromString(
    vendorServiceConfigText
  );

  if (!vendorServiceConfig.every(isVendorServiceConfigRow))
    throw new Error("Invalid vendor service config");

  return vendorServiceConfig;
})();

export const getVendorServiceConfigRow = async (
  fields: Partial<VendorServiceConfigRow>
): Promise<VendorServiceConfigRow> => {
  const vendorServiceConfig = await cachedVendorServiceConfig;

  const vendorServiceConfigRow = vendorServiceConfig.find((row) =>
    Object.entries(fields).every(
      ([fieldName, fieldValue]) =>
        fieldValue === undefined ||
        row[fieldName as keyof VendorServiceConfigRow] === fieldValue
    )
  );

  if (vendorServiceConfigRow === undefined)
    throw Error("Vendor service config row not found");

  return vendorServiceConfigRow;
};
```

That self invocation is a bit freaky so here's a model for this I played around with in a terminal to assure myself that `getS3Object` wouldn't be called over and over 
```
const slow = async (arg) =>
  new Promise((resolve) => {
    console.log('new underlying promise spawned')
    setTimeout(() => {
      resolve(`You sent me ${arg}`);
    }, 1000);
  });

const data = (async () => {
  const res = await slow('thing');
  return res;
})();

(async () => {
    const a = await data;
    const b = await data;
    const c = await data;
    console.log(a,b,c)
})()
```

File: integration_tests/helpers/configHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+const isVendorServiceConfigRow = (x: any): x is VendorServiceConfigRow =>
```
Comment (Si3rr4wow) [inline integration_tests/helpers/configHelpers.ts:54]: love a predicate

File: integration_tests/helpers/mock-data/invoice/random.ts [hunk 3]
```diff
@@ -53,37 +62,49 @@ const randomCustomer = (): Customer => {
...
-  const subtotal = quantity * priceTier.unitPrice;
+  const unitPrice = options?.unitPrice ?? priceTier.unitPrice;
+  const subtotal = options?.subtotal ?? quantity * unitPrice;
   return {
-    description: "Verification of sentience via address checking mechanism", // should match IPV_ADDRESS_CRI_END from fake-vendor-services.csv
-    vat: 0.2 * subtotal,
+    description:
+      options?.description ??
+      "Verification of sentience via address checking mechanism", // should match IPV_ADDRESS_CRI_END from fake-vendor-services.csv
+    vat: options?.vat ?? 0.2 * subtotal,
```
Comment (Si3rr4wow) [inline integration_tests/helpers/mock-data/invoice/random.ts:80]: maybe overkill but could we have the option to also pass in a VAT rate that replaces 0.2 if present?

File: integration_tests/helpers/mock-data/invoice/random.ts [hunk 4]
```diff
@@ -53,37 +62,49 @@ const randomCustomer = (): Customer => {
...
-    unitPrice: priceTier.unitPrice,
+    unitPrice,
     subtotal,
   };
 };
 
 const randomLineItems = (quantity: number): LineItem[] => {
-  return new Array(quantity).fill(null).map(randomLineItem);
+  return new Array(quantity).fill(null).map(() => randomLineItem());
 };
```
Comment (Si3rr4wow) [inline integration_tests/helpers/mock-data/invoice/random.ts:89]: any ideas for a slick way to pass overrides for individual line items to this function so we can control them from `randomInvoice`'s options?

**Review by MelAtGDS — APPROVED** [review]

**Pull request merged**